src/query/src/range_select/plan.rs

```diff
@@ -1141,29 +1152,47 @@ impl RangeSelectStream {
         // 1 for time index column
         let mut columns: Vec<Arc<dyn Array>> =
             Vec::with_capacity(1 + self.range_exec.len() + self.by.len());
-        let mut ts_builder = TimestampMillisecondBuilder::with_capacity(self.output_num_rows);
-        let mut all_scalar = vec![Vec::with_capacity(self.output_num_rows); self.range_exec.len()];
-        let mut by_rows = Vec::with_capacity(self.output_num_rows);
+        let mut ts_builder = TimestampMillisecondBuilder::with_capacity(self.num_not_null_rows);
+        let mut all_scalar =
+            vec![Vec::with_capacity(self.num_not_null_rows); self.range_exec.len()];
+        let mut by_rows = Vec::with_capacity(self.num_not_null_rows);
         let mut start_index = 0;
-        // RangePlan is calculated on a row basis. If a column uses the PREV or LINEAR filling strategy,
-        // we must arrange the data in the entire data row to determine the NULL filling value.
-        let need_sort_output = self
+        // If any range expr need fill, we need fill both the missing align_ts and null value.
+        let need_fill_output = self.range_exec.iter().any(|range| range.fill.is_some());
+        // The padding value for each accumulator
+        let padding_values = self
             .range_exec
             .iter()
-            .any(|range| range.fill == Fill::Linear || range.fill == Fill::Prev);
+            .map(|e| e.expr.create_accumulator()?.evaluate())
+            .collect::<DfResult<Vec<_>>>()?;
         for SeriesState {
             row,
             align_ts_accumulator,
         } in self.series_map.values()
         {
-            // collect data on time series
-            let mut align_ts = align_ts_accumulator.keys().copied().collect::<Vec<_>>();
-            if need_sort_output {
-                align_ts.sort();
+            // skip empty time series
+            if align_ts_accumulator.is_empty() {
+                continue;
             }
+            // find the first and last align_ts
```

Comment:
Can we add a fast path when `align_ts_accumulator` only has one element?

Reply:
In this scenario, many dispose logic still duplicate, and the readability of the code will decrease.